### PR TITLE
Fix collage section product and collection card image size

### LINKED ISSUE #995: [Global setting] Collage section product image are too small
### Observation 

- [ ] [Image for products](https://screenshot.click/21-12-gvcwr-fqys8.png) in the collage section is very small.

### PR DESCRIPTION
**Pull request by KaichenWang**
## **Why are these changes introduced?**

Fixes #995

Also addresses:

https://github.com/Shopify/dawn/issues/929#issuecomment-992755128
https://github.com/Shopify/dawn/issues/378
https://github.com/Shopify/dawn/issues/699
https://github.com/Shopify/dawn/issues/910

### 1. Prevent "squished" images

Product and collection card images:
- rendered in their native aspect ratio
- cannot be "squished" beyond minimum height (as determined by aspect ratio)
- can stretch vertically to fill up remaining vertical space (see point # 2)

This is a change in how the collage inherently behaves (see _before_ and _after_ demo stores), but will prevent narrow horizontally cropped product/collection images.

### 2. Extend card height

- Allow product and collection cards to extend height to fill available vertical container space
- Boolean `extend_height` can be passed into `card-product` and `card-collection` snippets

### 3. Refactor - Separate collage grid layout from block content

This allows styles for collage grid layout to be decoupled from each block's contents. This simplifies the CSS structure.

#### Layout
| element                | purpose                                                |
|----------------|-------------------------------------------------|
| `.collage`       | container grid                                  |
| `.collage__item` | grid items; containers for block content |

#### Block content
| block type | in Liquid                     |
|------------|-------------------------------|
| `image`     | `.collage-card` `.media`          |
| `product`    | `card-product` snippet          |
| `collection` | `card-collection` snippet       |
| `video`      | `.collage-card` `.deferred-media` |

### 4. Refactor - Separate video modal into its own component

- Video modal could be used outside of a collage context, so makes sense to decouple from collage.
- Simplify BEM naming convention

| before                            | after                          |
|-----------------------------------|--------------------------------|
| in `collage.css`                  | in `component-modal-video.css` |
| `.collage-video__modal`           | `.modal-video`                 |
| `.collage-video__modal-<element>` | `.modal-video__<element>`      |

#### Before
- If video is the only collage block, play video inline
- If video is not the only collage block, play video in modal

#### After
- Always play collage video in modal

#### Why?
- Was an instance of "magic", where it may be unclear to a merchant why the collage video has 2 different behaviours
- Removes significant repeated code in Liquid
- Thanks to vertical spacing settings, the existing Video section can now essentially be used in conjunction with a collage in order to achieve inline video playback

## **Checklist**
- [ ] Followed [theme code principles](https://github.com/Shopify/dawn/blob/main/.github/CONTRIBUTING.md#theme-code-principles)
- [ ] Linted with [Theme Check](https://github.com/Shopify/theme-check)
- [ ] Tested on [mobile](https://shopify.dev/themes/store/requirements#mobile-browser-requirements)
- [ ] Tested on [multiple browsers](https://shopify.dev/themes/store/requirements#desktop-browser-requirements)
- [ ] Tested for [accessibility](https://shopify.dev/themes/best-practices/accessibility)

### CODE REVIEW

**Comment by KaichenWang** [comment]
**Known issue I'm working on right now**: Collection cards with no image - Title is vertically misaligned

Screen Shot 2021-12-14 at 4 51 39 PM

**Review by melissaperreault — COMMENTED** [review]
### Global setting relationship feedback

**Image padding**
- [x] Remove the image padding from the `Collection block` and `Product block` and let the `Cards` global setting manage that. 
- [x] For Image and Video blocks, `Media` global setting don't offer a setting to manage padding. We could keep it for both to accomplish the expected aesthetic with this section. To be discussed!

**Color scheme**
- [x] As per IRL conversation, please let's remove the `Color scheme` setting at the `Collection bock` and `Product block` levels until we have a different color approach.


**Product block**
- [ ] Remove the `Show secondary background` setting as it would interfere with the global color scheme.

**Comment by wiktoriaswiecicka** [comment]
Collage is looking great. I've got one question - On the collection block card - Is it possible to apply rounded corners to all corners on the image when image padding is enabled? Right now it's only top corners:

Screen Shot 2021-12-15 at 5 35 40 PM

**Review by wiktoriaswiecicka — APPROVED** [review]

**Review by ludoboludo — COMMENTED** [review]
I only reviewed some of the functionality and have a few things I noted.

- [ ] The focused image used to dictate the overall height of the section and wouldn't crop the image/media in it unless it's very small. Seems like now the height is quite tall by default, I'd say too tall in general. If we're wanting to remove that previous functionality then it would probably make sense to add a section height setting like some other sections. We could have an adapt to focused block image as well 🤷  [screenshot](https://screenshot.click/21-12-oxctc-f0d31.png). 
This change would have an impact on how people used to use the section but I guess it would be part of a breaking change so it's not like we would update people to this version without their consent.

- [ ] A color scheme could be nice on the section level but I see Meli mentioned that it would be done after future explorations.

- [ ] The collection block title seem a little to big when there isn't an image in comparison to the section title and product block title. [I had started a PR](https://github.com/Shopify/dawn/pull/920) to fix it but might as well be done in this overall rework of the section. [screenshot](https://screenshot.click/21-12-ljzs5-pvjj4.png)

- [ ] The spacing on the product card that has no image seem a little too far away from the title with the border: [screenshot](https://screenshot.click/21-12-3bfbg-t5a5d.png). Maybe the space above it could reduced to visually convey that they're related. 

- [ ] In the context of this section, should the image and video block be considered "cards" ? I just find it odd when you're tweaking the cards' border and shadow, that it impacts only some blocks. So if you want the overall look of the section to be similar you have to tweak different stuff in the theme settings. I guess with potential section overrides it would make it work more easily. [screenshot](https://screenshot.click/21-12-lbqsk-k25k1.png)

- [ ] On mobile it looks like if you have two blocks the layout isn't what it should be: [screenshot](https://screenshot.click/21-12-sk6ft-03zw4.png). 

- [ ] There is this situation with collage where if you remove the title of the second collage section (when you have two of them in a row), it then looks like one section. So I think in this situation we should use the `multicolumn vertical space` instead of the `vertical space between sections` 🤔  [screenshot](https://screenshot.click/21-12-7c7wf-l73ls.png)

- [ ] On the video block, the shadow setting doesn't not seem to be setup though the border is working: [screenshot](https://screenshot.click/21-12-7bun7-55qz8.png)

**Comment by KaichenWang** [comment]
@ludoboludo 

Thanks for reviewing!

## I've addressed the following

> * [x]  On mobile it looks like if you have two blocks the layout isn't what it should be: [screenshot](https://screenshot.click/21-12-sk6ft-03zw4.png).

> * [x]  On the video block, the shadow setting doesn't not seem to be setup though the border is working: [screenshot](https://screenshot.click/21-12-7bun7-55qz8.png)

## Here's some more info on the other points

> * [ ]  The focused image used to dictate the overall height of the section and wouldn't crop the image/media in it unless it's very small. Seems like now the height is quite tall by default, I'd say too tall in general. If we're wanting to remove that previous functionality then it would probably make sense to add a section height setting like some other sections. We could have an adapt to focused block image as well 

Correct, the desktop grid item behaviour has changed

### Before

Screen Shot 2021-12-17 at 12 11 52 PM

### After

All collage blocks' images try to maintain their native aspect ratio, but will expand to fit any available vertical space. The key difference is that non-focus item images cannot shrink vertically anymore. This is helps maintain overall larger surface area for images in the collage. Maximizing surface area for images is important, because global settings like border, border-radius, image padding and large grid gaps can start to eat into the space for the image itself.

Screen Shot 2021-12-17 at 12 16 13 PM

> * [ ]  A color scheme could be nice on the section level but I see Meli mentioned that it would be done after future explorations.

Correct, this would be implemented as part of block-level overrides.

> * [ ]  The collection block title seem a little to big when there isn't an image in comparison to the section title and product block title. [I had started a PR](https://github.com/Shopify/dawn/pull/920) to fix it but might as well be done in this overall rework of the section. [screenshot](https://screenshot.click/21-12-ljzs5-pvjj4.png)

The collection block now shares all markup and styles with the `card-collection` component. I agree the font-size can be reduced here, but would prefer to make that change separately; to ensure more visibility for the change and proper testing globally.

> * [ ]  The spacing on the product card that has no image seem a little too far away from the title with the border: [screenshot](https://screenshot.click/21-12-3bfbg-t5a5d.png). Maybe the space above it could reduced to visually convey that they're related.

Same for the product block, which now shares all markup and styles with the `card-product` component. Would prefer to test the proposed change in more scenarios and ship separately.

> * [ ]  In the context of this section, should the image and video block be considered "cards" ? I just find it odd when you're tweaking the cards' border and shadow, that it impacts only some blocks. So if you want the overall look of the section to be similar you have to tweak different stuff in the theme settings. I guess with potential section overrides it would make it work more easily. [screenshot](https://screenshot.click/21-12-lbqsk-k25k1.png)

Yes, [separate issue was created for this](https://github.com/Shopify/dawn/issues/1015)

> * [ ]  There is this situation with collage where if you remove the title of the second collage section (when you have two of them in a row), it then looks like one section. So I think in this situation we should use the `multicolumn vertical space` instead of the `vertical space between sections` 🤔  [screenshot](https://screenshot.click/21-12-7c7wf-l73ls.png)

Yeah, although it could be beneficial for automatically enforcing consistent grid spacing, the decision was made to avoid "magic" behaviour. The section's [internal padding settings](https://github.com/Shopify/dawn/pull/1006) should be leveraged to achieve the same look.

**Comment by melissaperreault** [comment]
Thank you Kai! I think I wasn't looking at the right demo link, sorry! 🙌 

- Note FYI: The `Image block` would lose the color scheme once we _Cards-ify_ all Collage blocks with https://github.com/Shopify/dawn/issues/1015

**Review by tyleralsbury — APPROVED** [review]
Looks really good.  I'm loving the simplification of the collage CSS.  I think a lot of what we've done in the past few weeks has added some complexity to our styles in particular, but a PR like this helps to reel it in a bit.  🎉 

I did some tests on various layouts and everything looked good to me.  I will trust @wiktoriaswiecicka 's certainly more thorough visual tests.

File: assets/collage.css [hunk 1]
```diff
@@ -5,492 +5,157 @@
...
+  .collage--mobile {
+    grid-template-columns: repeat(2, 1fr);
   }
 
-  .collage-card--right:nth-child(2),
-  .collage-card--right:nth-child(3) {
-    grid-column: 2 / span 2;
+  .collage--mobile .collage__item--left:nth-child(3n - 2),
+  .collage--mobile .collage__item--left:nth-child(3n - 1):last-child {
+    grid-column: span 2;
```
[the remark this answers is not in the document]
Comment (KaichenWang) [reply assets/collage.css:25]: Yes good point! Updated https://github.com/Shopify/dawn/pull/1014/commits/44485c5392c4d71df1508c2f6f9600698ec4d6ba

**Pull request merged**